### PR TITLE
Build the vsix as part of the default build cycle

### LINKED ISSUE #1029: Make the VSIX a part of the default build output
Once our build agents have VS 2017 installed, we should be able to do this.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #1029

### CODE REVIEW

File: build/VSIX.targets [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    <ItemGroup>
+      <MSBuild15ExePaths Include="$(MSBuildProgramFiles32)\Microsoft Visual Studio\2017*\**\MSBuild\15.0\Bin\MSBuild.exe" />
+    </ItemGroup>
+
+    <Warning
+      Text="Unable to locate MSBuild 15.0 under $(MSBuildProgramFiles32)\Microsoft Visual Studio"
+      Condition="'@(MSBuild15ExePaths)'==''"/>
+
+    <PropertyGroup Condition="'@(MSBuild15ExePaths)'!=''">
+      <MSBuildExePath>%(MSBuild15ExePaths.FullPath)</MSBuildExePath>
```
Comment (dougbu) [inline build/VSIX.targets]: Anyone with multiple VS 2017 instances installed will get more than one path in `$(MSBuildExePath)`. Does the `BuildVSIX` target work on such a machine?

File: build/VSIX.targets [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    <NuGetCommandLineVersion>3.5.0</NuGetCommandLineVersion>
+    <VSIXName>Microsoft.VisualStudio.RazorExtension</VSIXName>
+    <VSIXProject>$(RepositoryRoot)tooling\$(VSIXName)\$(VSIXName).csproj</VSIXProject>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="NuGet.CommandLine" Version="$(NuGetCommandLineVersion)" />
+  </ItemGroup>
+
+  <Target Name="GenerateVSIX" DependsOnTargets="LocateMSBuildExe;BuildVSIX" />
```
Comment (dougbu) [inline build/VSIX.targets]: All targets except this one should follow @natemcmaster's internal convention -- start name with an underscore. Nothing else can usefully be called except from within this file.

File: build/VSIX.targets [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  <Target Name="RestoreProject">
+    <PropertyGroup>
+      <NuGetExe>$(NuGetPackageFolders)nuget.commandline\$(NuGetCommandLineVersion)\tools\nuget.exe</NuGetExe>
+    </PropertyGroup>
+    <Exec Command="$(NuGetExe) restore &quot;$(VSIXProject)&quot; -SolutionDir &quot;$(RepositoryRoot) &quot; -verbosity quiet" />
+  </Target>
+
+  <Target Name="BuildVSIX"
+    DependsOnTargets="LocateMSBuildExe"
+    Condition="'$(MSBuildExePath)'!=''">
```
Comment (dougbu) [inline build/VSIX.targets]: Remove the `DependsOnTargets` attribute because it implies `BuildVSIX` can work correctly standalone.

File: build/VSIX.targets [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    <PropertyGroup Condition="'@(MSBuild15ExePaths)'!=''">
+      <MSBuildExePath>%(MSBuild15ExePaths.FullPath)</MSBuildExePath>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="RestoreProject">
+    <PropertyGroup>
+      <NuGetExe>$(NuGetPackageFolders)nuget.commandline\$(NuGetCommandLineVersion)\tools\nuget.exe</NuGetExe>
+    </PropertyGroup>
+    <Exec Command="$(NuGetExe) restore &quot;$(VSIXProject)&quot; -SolutionDir &quot;$(RepositoryRoot) &quot; -verbosity quiet" />
```
Comment (dougbu) [inline build/VSIX.targets]: Call the `Restore` target instead; shouldn't need to `Exec`.

File: build/VSIX.targets [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      <VSIXOutputPath>$(BuildDir)$(VSIXName).vsix</VSIXOutputPath>
+    </PropertyGroup>
+
+    <CallTarget Targets="RestoreProject" />
+    <ItemGroup>
+      <MSBuildArguments Include="
+        $(VSIXProject);
+        /v:M;
+        /fl;
+        /flp:v=D;
```
Comment (dougbu) [inline build/VSIX.targets]: Detailed is the default verbosity level. Leave this out.

File: build/VSIX.targets [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        /flp:LogFile=$(VSIXLogFilePath);
+        /p:DeployExtension=false;
+        /p:TargetVSIXContainer=$(VSIXOutputPath);
+        /p:Configuration=$(Configuration);" />
```
Comment (dougbu) [inline build/VSIX.targets]: No need for the trailing semicolons. Newlines keep the arguments separate.

File: build/repo.targets [hunk 7]
```diff
@@ -1,12 +1,10 @@
...
-    The .NET Core SDK doesn't currently support Microsoft.VsSDK.targets.
-  -->
+  <Import Project="VSIX.targets" />
   <ItemGroup>
     <Solutions Update="..\Razor.sln">
       <!-- the 'DebugNoVSIX' and 'ReleaseNoVSIX' configurations exclude the VSIX project, which doesn't build with Microsoft.NET.Sdk yet. -->
       <AdditionalProperties>Configuration=$(Configuration)NoVSIX</AdditionalProperties>
     </Solutions>
   </ItemGroup>
+
```
Comment (dougbu) [inline build/repo.targets]: 👎 doesn't make the file any easier to read

File: tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest [hunk 8]
```diff
@@ -2,7 +2,7 @@
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
   <Metadata>
     <Identity Id="Microsoft.VisualStudio.RazorExtension" Version="0.1" Language="en-US" Publisher="Microsoft" />
-    <DisplayName>Microsoft.VisualStudio.RazorExtension</DisplayName>
+    <DisplayName>Razor Lanaguage Services</DisplayName>
```
Comment (dougbu) [inline tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest]: What does this have to do with the rest of the PR?

File: build/repo.targets [hunk 9]
```diff
@@ -1,12 +1,10 @@
 <Project>
-  <!--
-    TODO include VSIX build in default lifecycle builds
-    The .NET Core SDK doesn't currently support Microsoft.VsSDK.targets.
```
Comment (dougbu) [inline build/repo.targets]: Side question: Do we have other checked-in TODO items in this repo?

File: build/VSIX.targets [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      <VSIXLogFilePath>$(ArtifactsDir)msbuild\vsix.log</VSIXLogFilePath>
+      <VSIXResponseFilePath>$(ArtifactsDir)msbuild\vsix.rsp</VSIXResponseFilePath>
+      <VSIXOutputPath>$(BuildDir)$(VSIXName).vsix</VSIXOutputPath>
+    </PropertyGroup>
+
+    <CallTarget Targets="RestoreProject" />
+    <ItemGroup>
+      <MSBuildArguments Include="
+        $(VSIXProject);
+        /v:M;
```
Comment (dougbu) [inline build/VSIX.targets]: Spell out `/verbosity:minimal` so future readers don't need to look this up.

File: build/VSIX.targets [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        /p:TargetVSIXContainer=$(VSIXOutputPath);
+        /p:Configuration=$(Configuration);" />
+    </ItemGroup>
+
+    <WriteLinesToFile
+      File="$(VSIXResponseFilePath)"
+      Lines="@(MSBuildArguments)"
+      Overwrite="true" />
+
+    <Exec Command="&quot;$(MSBuildExePath)&quot; @&quot;$(VSIXResponseFilePath)&quot;" />
```
Comment (dougbu) [inline build/VSIX.targets]: Do we have a bug tracking the possibility `dotnet msbuild` supports VSIX creation (at least on Windows)?

File: tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest (hunk 8 above)
Comment (natemcmaster) [inline tooling/Microsoft.VisualStudio.RazorExtension/source.extension.vsixmanifest]: Typo: Lanaguage => Language

File: build/VSIX.targets (hunk 3 above)
Comment (natemcmaster) [inline build/VSIX.targets]: ?? I don't follow.

File: build/VSIX.targets (hunk 1 above)
Comment (natemcmaster) [inline build/VSIX.targets]: You can use `<FindInList>` to filter to the first one found.

File: build/VSIX.targets [hunk 12]
```diff
@@ -0,0 +1,67 @@
+<Project>
+  <PropertyGroup>
+    <PackageDependsOn>$(PackageDependsOn);GenerateVSIX</PackageDependsOn>
+    <NuGetCommandLineVersion>3.5.0</NuGetCommandLineVersion>
+    <VSIXName>Microsoft.VisualStudio.RazorExtension</VSIXName>
+    <VSIXProject>$(RepositoryRoot)tooling\$(VSIXName)\$(VSIXName).csproj</VSIXProject>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="NuGet.CommandLine" Version="$(NuGetCommandLineVersion)" />
```
Comment (natemcmaster) [inline build/VSIX.targets]: Can be conditional on OS == Windows_NT.

File: build/VSIX.targets (hunk 6 above)
Comment (natemcmaster) [inline build/VSIX.targets]: They're required. This is an item group, not a property.

File: build/VSIX.targets [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  </Target>
+
+  <Target Name="RestoreProject">
+    <PropertyGroup>
+      <NuGetExe>$(NuGetPackageFolders)nuget.commandline\$(NuGetCommandLineVersion)\tools\nuget.exe</NuGetExe>
+    </PropertyGroup>
+    <Exec Command="$(NuGetExe) restore &quot;$(VSIXProject)&quot; -SolutionDir &quot;$(RepositoryRoot) &quot; -verbosity quiet" />
+  </Target>
+
+  <Target Name="BuildVSIX"
```
Comment (natemcmaster) [inline build/VSIX.targets]: Suggest also making this conditional on OS == Windows_NT.

File: build/VSIX.targets (hunk 4 above)
Comment (natemcmaster) [inline build/VSIX.targets]: Actually this is correct. The restore target doesn't know how to restore NuGet 2 (packages.config).

File: build/VSIX.targets (hunk 1 above)
Comment (pranavkm) [inline build/VSIX.targets]: This is fewer lines of code and works well. Unless you have a strong preference for `FindInList`, I'll leave this in.

Comment (natemcmaster) [inline build/VSIX.targets]: Nvm, didn't read carefully. Yes, this code is fine as-is.

File: build/VSIX.targets (hunk 10 above)
Comment (natemcmaster) [inline build/VSIX.targets]: abbr ok w/ me too.

**Comment by pranavkm** [comment]
🆙 :date

File: build/VSIX.targets [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    <VSIXProject>$(RepositoryRoot)tooling\$(VSIXName)\$(VSIXName).csproj</VSIXProject>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="NuGet.CommandLine" Version="$(NuGetCommandLineVersion)" Condition="'$(OS)'=='Windows_NT'" />
+  </ItemGroup>
+
+  <Target
+    Name="GenerateVSIX"
+    DependsOnTargets="_LocateMSBuildExe;_BuildVSIX" />
```
Comment (dougbu) [inline build/VSIX.targets]: Probably should make this target also conditional on `'$(OS)'=='Windows_NT'`.

File: build/VSIX.targets [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    <Warning
+      Text="Unable to locate MSBuild 15.0 under $(MSBuildProgramFiles32)\Microsoft Visual Studio"
+      Condition="'@(MSBuild15ExePaths)'==''"/>
+
+    <PropertyGroup Condition="'@(MSBuild15ExePaths)'!=''">
+      <MSBuildExePath>%(MSBuild15ExePaths.FullPath)</MSBuildExePath>
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="RestoreProject">
```
Comment (dougbu) [inline build/VSIX.targets]: Shouldn't this also be effectively-internal i.e. `_RestoreProject`?
